library/ssl_tls.c

```diff
@@ -711,7 +711,7 @@ static int ssl_handshake_init( mbedtls_ssl_context *ssl )
     if( ssl->transform_negotiate )
         mbedtls_ssl_transform_free( ssl->transform_negotiate );
     if( ssl->session_negotiate )
-        mbedtls_ssl_session_free( ssl->session_negotiate );
+        mbedtls_ssl_session_free_chain( ssl->session_negotiate );
```

Comment:
I think in `mbedtls_ssl_free` there is also an `mbedtls_ssl_session_free( ssl->session_negotiate )` that needs to be changed to an `mbedtls_ssl_session_free_chain( ssl->session_negotiate )`.